File: Graphs/24/topological_sort.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function (numCourses, prerequisites) {
  const adjList = new Array(numCourses).fill(0).map(() => []);
  const indegree = new Array(numCourses).fill(0);
  for (const item of prerequisites) {
    const [child, parent] = item;
    adjList[parent].push(child);
    indegree[child]++;
  }
  let num = 0;
  const stack = [];

  for (let v = 0; v < numCourses; v++) {
    if (indegree[v] === 0) {
      stack.push(v);
    }
  }

  while (stack.length) {
    const v = stack.pop();
    num++;
    for (const node of adjList[v]) {
      indegree[node]--;
      if (indegree[node] === 0) {
        stack.push(node);
      }
    }
  }

  return num === numCourses ? true : false;
};

File: Graphs/24/topological_sort.py
from collections import deque


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adjList = [[] for _ in range(numCourses)]
        indegree = [0]*numCourses
        for item in prerequisites:
            [child, parent] = item
            adjList[parent].append(child)
            indegree[child] += 1
        num = 0
        stack = []

        for v in range(numCourses):
            if indegree[v] == 0:
                stack.append(v)

        while len(stack):
            v = stack.pop()
            num += 1
            for node in adjList[v]:
                indegree[node] -= 1
                if indegree[node] == 0:
                    stack.append(node)
        
        return True if num == numCourses else False
# Time : O(n)
# Space: O(n^2)